synapse/handlers/receipts.py

```diff
@@ -189,7 +189,7 @@ def __init__(self, hs: "HomeServer"):
 
     @staticmethod
     def filter_out_private_receipts(
-        rooms: List[JsonDict], user_id: str
+        rooms: Sequence[JsonDict], user_id: str
```

Comment:
General comment: I wonder if we ought to use `JsonMapping` more (The sequence may be immutable, but I can still use it to get references to mutable data and e.g. pop a key-value pair.)

For cached methods specifically, this is the kind of thing that https://github.com/matrix-org/synapse/pull/14911 can help with in the future

Reply:
I think so

I suggest that we:

1. Don't do that here.
2. Rebase #14911 after merging this.
3. Fix-up interior values to be non-mutable (including using `JsonMapping`).